### ISSUE TITLE
cljdoc build failed

### ISSUE THREAD

**Issue by behrica** [scicloj/clojuress#94]
I added cljdoc badge for Clojisr.
Build is failing:
https://cljdoc.org/builds/78095

Not sure, if easy fix

**Comment by genmeblog** [comment]
`cljdoc` tries to load a namespace and run a R session

```
INFO: [:clojisr.v1.session/make-session {:action :new-session, :id nil, :actual-session-args {:session-type :rserve}}]
{:clojure.main/message
 "Execution error (AssertionError) at clojisr.v1.impl.rserve.proc/r-path (proc.clj:30).\nAssert failed: (not= % \"\")\n",
 :clojure.main/triage
 {:clojure.error/class java.lang.AssertionError,
  :clojure.error/line 30,
  :clojure.error/cause "Assert failed: (not= % \"\")",
  :clojure.error/symbol clojisr.v1.impl.rserve.proc/r-path,
  :clojure.error/source "proc.clj",
  :clojure.error/phase :execution},
 :clojure.main/trace
```

**Comment by behrica** [comment]
Yes,  noticed it.
This line https://github.com/scicloj/clojisr/blob/82b4b77e53cc938c95694f19512c6b0187de3a5b/src/clojisr/v1/applications/plotting.clj#L14

fails on CI, correct ?

**Comment by behrica** [comment]
I will try to add "no-doc" metadata:
https://github.com/cljdoc/cljdoc/blob/master/doc/userguide/for-library-authors.adoc#declaring-your-public-api
Not sure, if it prevents ns loading.

**Comment by behrica** [comment]
It would only be visible on a new release, I believe.
cljdocs build from maven jars, not from source.
I propose we wait for next release and see if the cljdoc build works.

Its a nice way to have clojure docs for the API, so we should support it, I think

**Comment by genmeblog** [comment]
Yes, it fails on CI. There is no R in the environment. `clojisr.v1.r` also will try to run a backend.

https://github.com/scicloj/clojisr/blob/82b4b77e53cc938c95694f19512c6b0187de3a5b/src/clojisr/v1/r.clj#L127

**Comment by behrica** [comment]
I thought so....

**Comment by behrica** [comment]
Maybe its overkill just to solve this issue,
but did you ever try to do these "def" lazy and via "intern"..

**Comment by genmeblog** [comment]
We were working on new session management system to solve this issue. But never finished.
Maybe there is a way to refactor current setup without changing the API but I don't know it now. Any concrete ideas?

**Comment by behrica** [comment]
I got this working:
```
(defn r== [e1 e2] ((clojisr.v1.r/r "`==`") e1 e2))
```

This would avoid the `def` and so the ns loading should "do nothing".
Not sure, this can be done for "all"   def in the ns.

**Comment by genmeblog** [comment]
This way every call `(r "`==`")` will be reevaluated, RObject created and so on.

Anyway this is very good idea and I like it. We also get a function instead of var.

**Comment by behrica** [comment]
I tried as well for `r$`, works in same way.
Are all these defs about "binary" functions ?
Or as well functions taking any number of args ?

**Comment by genmeblog** [comment]
I don't know now. We have to go one by one.
This strategy can be used also for plotting by replacing all RObjects calls into a string representation.

`r.grDevices/dev-off` -> "dev.off()" etc...

**Comment by behrica** [comment]
Not sure, it can fully work for Clojisr, but it would follow a "clojure principle" that ns loading should not do "anything".

**Comment by behrica** [comment]
I make a new issue at least , and link to this one.

**Comment by behrica** [comment]
> I don't know now. We have to go one by one. This strategy can be used also for plotting by replacing all RObjects calls into a string representation.
> 
> `r.grDevices/dev-off` -> "dev.off()" etc...

Would this be needed ?
Or just do the " (require-r '[grDevices]) " later , "inside" the plot function ?
Maybe with a `delay` to avoid repeated calls ?

**Comment by genmeblog** [comment]
> > I don't know now. We have to go one by one. This strategy can be used also for plotting by replacing all RObjects calls into a string representation.
> > `r.grDevices/dev-off` -> "dev.off()" etc...
> 
> Would this be needed ? Or just do the " (require-r '[grDevices]) " later , "inside" the plot function ? Maybe with a `delay` to avoid repeated calls ?

Yes, this is needed because later the symbol interned by `require-r` is used.

**Comment by genmeblog** [comment]
It should work now, but it doesn't yet. I think we have to hide `vrksasana` somewhere.

```
{:via
 [{:type clojure.lang.ExceptionInfo,
   :message
   "Could not generate Clojure documentation for vrksasana.impl.r.astgen",
   :data {},
   :at
   [cljdoc_analyzer.metagetta.utils$default_exception_handler
    invokeStatic
    "utils.clj"
    125]}
  {:type clojure.lang.Compiler$CompilerException,
   :message
   "Syntax error macroexpanding at (vrksasana/impl/r/astgen.clj:1:1).",
   :data
   #:clojure.error{:phase :execution,
                   :line 1,
                   :column 1,
                   :source "vrksasana/impl/r/astgen.clj"},
   :at [clojure.lang.Compiler load "Compiler.java" 8235]}
  {:type java.io.FileNotFoundException,
   :message
   "Could not locate tech/v3/protocols/dataset__init.class, tech/v3/protocols/dataset.clj or tech/v3/protocols/dataset.cljc on classpath.",
   :at [clojure.lang.RT load "RT.java" 482]}],
```

**Comment by genmeblog** [comment]
BTW we need to document functions :) Most of them are without docstrings.
